### PR TITLE
CMake-based build system

### PR DESCRIPTION
**Pull request by matthiasdiener**
*Original date: 2018-08-30 16:54:54*
*Original PR: https://charm.cs.illinois.edu/gerrit/4418*

---
Can build charm++, AMPI, and LIBS. Passes `make test` in `tests/` and `examples/`.

Tested on:
    - netlrts-{darwin,linux}-{x86_64,ppc} {,smp}
    - multicore-{darwin,linux}-x86_64
    - mpi-linux-x86_64 {,smp}
    - verbs-linux-x86_64 {,smp}
    - ofi-linux-x86_64
    - ucx-linux-x86_64

For MacOS 10.14+, you will likely need the header files installed in /usr/include: https://donatstudios.com/MojaveMissingHeaderFiles

A simple way to build with cmake is to replace your `./build` command with `./buildcmake`, which supports many of the options that `./build` supports:

``` bash
$ ./buildcmake AMPI netlrts-linux-x86_64 --with-production -j4
```

Note that this does not replace or change the existing build system, so the existing ways to build charm++ should not be affected.

### CODE REVIEW

**Comment by matthiasdiener** [comment]
This is ready for more testing & reviews.

**Comment by matthiasdiener** [comment]
@ericjbohm Could you please check the current version? The issues you mentioned should hopefully be fixed now.

**Comment by ericjbohm** [comment]
Problem with ROMIO during AMPI build remains. Also, -j or -j4 crash much earlier as they fail to handle dependencies correctly.  so this is with no -j, a sequential build.
./buildcmake LIBS netlrts-darwin-x86_64 --with-production
.
.
.

[ 99%] Performing download step (DIR copy) for 'romio'
[100%] No patch step for 'romio'
[100%] No update step for 'romio'
[100%] Performing configure step for 'romio'
configure: WARNING: Unknown architecture ... proceeding anyway
configure: WARNING: File locks may not work with NFS.  See the Installation and
users manual for instructions on testing and if necessary fixing this
[100%] Performing build step for 'romio'
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: /Users/ebohm/ppl/github/mastercharm/charm/netlrts-darwin-x86_64/src/libs/ck-libs/ampi/romio-prefix/src/romio/lib/libmpio.a(mpich2_fileutil.o) has no symbols
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: /Users/ebohm/ppl/github/mastercharm/charm/netlrts-darwin-x86_64/src/libs/ck-libs/ampi/romio-prefix/src/romio/lib/libmpio.a(mpich2_fileutil.o) has no symbols
malloc.c:24:10: fatal error: 'malloc.h' file not found
#include <malloc.h>
         ^~~~~~~~~~
1 error generated.
Fatal Error by charmc in directory /Users/ebohm/ppl/github/mastercharm/charm/netlrts-darwin-x86_64/src/libs/ck-libs/ampi/romio-prefix/src/romio/adio/common

**Comment by matthiasdiener** [comment]
> Problem with ROMIO during AMPI build remains. Also, -j or -j4 crash much earlier as they fail to handle dependencies correctly. so this is with no -j, a sequential build.
> ./buildcmake LIBS netlrts-darwin-x86_64 --with-production
> [ ... ]
> malloc.c:24:10: fatal error: 'malloc.h' file not found
> #include <malloc.h>
> ^~~~~~~~~~
> 1 error generated.
> Fatal Error by charmc in directory /Users/ebohm/ppl/github/mastercharm/charm/netlrts-darwin-x86_64/src/libs/ck-libs/ampi/romio-prefix/src/romio/adio/common

On which machine is this?

**Comment by ericjbohm** [comment]
> > Problem with ROMIO during AMPI build remains. Also, -j or -j4 crash much earlier as they fail to handle dependencies correctly. so this is with no -j, a sequential build.
> > ./buildcmake LIBS netlrts-darwin-x86_64 --with-production
> > [ ... ]
> > malloc.c:24:10: fatal error: 'malloc.h' file not found
> > #include <malloc.h>
> > ^~~~~~~~~~
> > 1 error generated.
> > Fatal Error by charmc in directory /Users/ebohm/ppl/github/mastercharm/charm/netlrts-darwin-x86_64/src/libs/ck-libs/ampi/romio-prefix/src/romio/adio/common
> 
> On which machine is this?

MacOS 10.13.6 Macbook pro.

**Comment by matthiasdiener** [comment]
> MacOS 10.13.6 Macbook pro.

The malloc.h ROMIO failure should be fixed now. I can't reproduce the build failure when running with `-j4`. Can you post the error that you get?

**Comment by ericjbohm** [comment]
Seems to have addressed the -j4 issues, but now its failing to pass the privatization test.

Its a little hard to follow where exactly the test falls over

cd privatization/
Erics-MacBook-Pro-4:privatization ebohm$ make test | tee foo
../../../bin/testrun  ./cxx-tlsglobals +p1 +vp2  
Charmrun> scalable start enabled. 
Charmrun> started all node programs in 1.114 seconds.
Charm++> Running in non-SMP mode: 1 processes (PEs)
Converse/Charm++ Commit ID: 6.9.0
Charm++> scheduler running in netpoll mode.
CharmLB> Load balancer assumes all CPUs are same.
Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
Charm++> cpu topology info is gathered in 0.000 seconds.
CharmLB> RotateLB created.
Beginning round of testing.
.
.
.
Testing: extern global, in same object
------------- Processor 0 Exiting: Called CmiAbort ------------
Reason: AMPI: Application called MPI_Abort()! 
.
.
.
Round of testing complete.
48 tests failed.

**Comment by matthiasdiener** [comment]
> Seems to have addressed the -j4 issues, but now its failing to pass the privatization test.
> 
> Its a little hard to follow where exactly the test falls over
> 
> cd privatization/
> Erics-MacBook-Pro-4:privatization ebohm$ make test | tee foo
> ../../../bin/testrun ./cxx-tlsglobals +p1 +vp2
> Charmrun> scalable start enabled.
> Charmrun> started all node programs in 1.114 seconds.
> Charm++> Running in non-SMP mode: 1 processes (PEs)
> Converse/Charm++ Commit ID: 6.9.0
> Charm++> scheduler running in netpoll mode.
> CharmLB> Load balancer assumes all CPUs are same.
> Charm++> Running on 1 hosts (1 sockets x 4 cores x 2 PUs = 8-way SMP)
> Charm++> cpu topology info is gathered in 0.000 seconds.
> CharmLB> RotateLB created.
> Beginning round of testing.
> .
> .
> .
> Testing: extern global, in same object
> ------------- Processor 0 Exiting: Called CmiAbort ------------
> Reason: AMPI: Application called MPI_Abort()!
> .
> .
> .
> Round of testing complete.
> 48 tests failed.
> 
> Charmrun> error on request socket to node 0 'localhost'--
> Socket closed before recv.
> make: *** [test-cxx-tlsglobals] Error 1

I can reproduce the failure with my build as well. @evan-charmworks do you have an idea what could be the issue? As far as I can see, the commands to build the test is exactly the same as when using the old build system

**Comment by evan-charmworks** [comment]
I am not able to build on Darwin.

`./buildcmake charm++ netlrts-darwin-x86_64`
```
Fatal Error by charmc in directory /Users/evan/charm/netlrts-darwin-x86_64
   file with unrecognized extension /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk
```

On Linux I get this:

```
$ cd tests/ampi/privatization/
$ make OPTS="-g3"
No privatization methods supported for current AMPI target.
```

**Comment by matthiasdiener** [comment]
> I am not able to build on Darwin.
> 
> `./buildcmake charm++ netlrts-darwin-x86_64`
> 
> ```
> Fatal Error by charmc in directory /Users/evan/charm/netlrts-darwin-x86_64
>    file with unrecognized extension /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk
> ```

I think this happens if you have a new xcode version (which doesn't install headers in /usr/include). I think I fixed that by following this procedure: http://icdevin.blogspot.com/2018/06/fixing-missing-headers-in-macos-mojave.html

**Comment by matthiasdiener** [comment]
> On Linux I get this:
> 
> ```
> $ cd tests/ampi/privatization/
> $ make OPTS="-g3"
> No privatization methods supported for current AMPI target.
> ```

This failure should be fixed now.

**Comment by matthiasdiener** [comment]
The AMPI privatization test should be fixed now.

**Comment by evan-charmworks** [comment]
Is there any reason why the minimum CMake version is 3.11? I had to set it to 3.10 to test on Linux and had no issues.

**Comment by matthiasdiener** [comment]
> Is there any reason why the minimum CMake version is 3.11? I had to set it to 3.10 to test on Linux and had no issues.

Line 276 in the main CMakeLists.txt is the only reason we require 3.11. Unless you are using Fortran, lower versions probably won't be an issue. There is a potential workaround outlined in a comment there which could reduce requirements to 3.6 (or even lower).

**Comment by matthiasdiener** [comment]
This needs another review.

**Comment by nitbhat** [comment]
I'm trying to test this patch with a darwin build on my mac. 

The build works with the previous build system `./build charm++ netlrts-darwin-x86_64 --with-production -j8`, but fails with`./buildcmake charm++ netlrts-darwin-x86_64 --with-production --suffix=cmake -j8` because of the following error:

```-- Check for working Fortran compiler: /usr/local/bin/gfortran
-- Check for working Fortran compiler: /usr/local/bin/gfortran  -- broken
CMake Error at /usr/local/Cellar/cmake/3.15.0/share/cmake/Modules/CMakeTestFortranCompiler.cmake:45 (message):
  The Fortran compiler

    "/usr/local/bin/gfortran"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /Users/nitinbhat/Work/software/charm/netlrts-darwin-x86_64/CMakeFiles/CMakeTmp

    Run Build Command(s):/usr/bin/make cmTC_47783/fast && /Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/cmTC_47783.dir/build.make CMakeFiles/cmTC_47783.dir/build
    Building Fortran object CMakeFiles/cmTC_47783.dir/testFortranCompiler.f.o
    /usr/local/bin/gfortran   -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk   -c /Users/nitinbhat/Work/software/charm/netlrts-darwin-x86_64/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_47783.dir/testFortranCompiler.f.o
    Linking Fortran executable cmTC_47783
    /usr/local/Cellar/cmake/3.15.0/bin/cmake -E cmake_link_script CMakeFiles/cmTC_47783.dir/link.txt --verbose=1
    /usr/local/bin/gfortran      -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk CMakeFiles/cmTC_47783.dir/testFortranCompiler.f.o  -o cmTC_47783
    collect2: fatal error: /usr/local/bin/gnm returned 1 exit status
    compilation terminated.
    make[1]: *** [cmTC_47783] Error 1
    make[1]: *** Deleting file `cmTC_47783'
    make: *** [cmTC_47783/fast] Error 2





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:28 (enable_language)


-- Configuring incomplete, errors occurred!
See also "/Users/nitinbhat/Work/software/charm/netlrts-darwin-x86_64/CMakeFiles/CMakeOutput.log".
See also "/Users/nitinbhat/Work/software/charm/netlrts-darwin-x86_64/CMakeFiles/CMakeError.log".```


Also, it looks like `--suffix=` is not supported.

**Comment by matthiasdiener** [comment]
--suffix is now supported. The build issue seems to be a known problem with home-brew on MacOS 10.14: https://github.com/Homebrew/homebrew-core/issues/32839

**Comment by nitbhat** [comment]
In the previous (regular) build system, the directory structure of tests/benchmarks/examples was copied inside <build-arch> with files being symlinked to the actual files. However, this is missing in the cmake build system support. 

I think it is worth adding this in order to maintain consistency with the previous build system and also to allow users/developers to work with tests/benchmarks/examples from multiple build-archs at the same time.

**Comment by nitbhat** [comment]
Tested `netlrts-darwin-x86_64` from scratch and with compilation after changes to source files. Both cases work as intended.

**Comment by nitbhat** [comment]
Build command: `./buildcmake LIBS ucx-linux-x86_64 --enable-error-checking --suffix=cmake --with-production -j16 |& tee result` on `iForge`. 

Seeing the following error related to permissions: 
```
[  6%] Building CXX object src/xlat-i/CMakeFiles/charmxi.dir/sdag/constructs/When.C.o
[  6%] Building CXX object src/xlat-i/CMakeFiles/charmxi.dir/sdag/constructs/While.C.o
[  6%] Linking CXX executable ../../bin/charmxi
[  6%] Built target charmxi
[  7%] Performing build step for 'hwloc'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Making all in src
  CC     topology.lo
  CC     traversal.lo
  CC     distances.lo
  CC     components.lo
  CC     bind.lo
  CC     bitmap.lo
  CC     pci-common.lo
  CC     diff.lo
  CC     misc.lo
  CC     base64.lo
  CC     topology-noos.lo
  CC     topology-synthetic.lo
  CC     topology-custom.lo
  CC     topology-xml.lo
  CC     topology-xml-nolibxml.lo
  CC     topology-linux.lo
  CC     topology-hardwired.lo
  CC     topology-x86.lo
  CCLD   libhwloc_embedded.la
Making all in include
[  7%] Performing install step for 'hwloc'
Making install in src
/bin/mkdir: cannot create directory `/usr/local/lib/hwloc': Permission denied
make[5]: *** [Makefile:654: install-pluginsLTLIBRARIES] Error 1
make[5]: *** Waiting for unfinished jobs....
make[4]: *** [Makefile:949: install-am] Error 2
make[3]: *** [Makefile:498: install-recursive] Error 1
make[2]: *** [CMakeFiles/hwloc.dir/build.make:74: hwloc-prefix/src/hwloc-stamp/hwloc-install] Error 2
make[1]: *** [CMakeFiles/Makefile2:68: CMakeFiles/hwloc.dir/all] Error 2
```

**Comment by matthiasdiener** [comment]
> Build command: `./buildcmake LIBS ucx-linux-x86_64 --enable-error-checking --suffix=cmake --with-production -j16 |& tee result` on `iForge`.
> 
> Seeing the following error related to permissions:
> [...]

Should be fixed now.

**Comment by nitbhat** [comment]
Seeing this error after that last patch:
```
[  6%] Linking CXX executable ../../bin/charmxi
[  6%] Built target charmxi
[  7%] Performing build step for 'hwloc'
make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
Making all in src
  CC     topology.lo
  CC     traversal.lo
  CC     distances.lo
  CC     components.lo
  CC     bind.lo
  CC     bitmap.lo
  CC     pci-common.lo
  CC     diff.lo
  CC     misc.lo
  CC     base64.lo
  CC     topology-noos.lo
  CC     topology-synthetic.lo
  CC     topology-custom.lo
  CC     topology-xml.lo
  CC     topology-xml-nolibxml.lo
  CC     topology-linux.lo
  CC     topology-hardwired.lo
  CC     topology-x86.lo
  CCLD   libhwloc_embedded.la
Making all in include
[  7%] No install step for 'hwloc'
[  7%] Completed 'hwloc'
[  7%] Built target hwloc
make: *** [Makefile:130: all] Error 2
```

**Comment by matthiasdiener** [comment]
> Seeing this error after that last patch:
> 
> ```
> [  6%] Linking CXX executable ../../bin/charmxi
> [  6%] Built target charmxi
> [  7%] Performing build step for 'hwloc'
> make[3]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
> Making all in src
>   CC     topology.lo
>   CC     traversal.lo
>   CC     distances.lo
>   CC     components.lo
>   CC     bind.lo
>   CC     bitmap.lo
>   CC     pci-common.lo
>   CC     diff.lo
>   CC     misc.lo
>   CC     base64.lo
>   CC     topology-noos.lo
>   CC     topology-synthetic.lo
>   CC     topology-custom.lo
>   CC     topology-xml.lo
>   CC     topology-xml-nolibxml.lo
>   CC     topology-linux.lo
>   CC     topology-hardwired.lo
>   CC     topology-x86.lo
>   CCLD   libhwloc_embedded.la
> Making all in include
> [  7%] No install step for 'hwloc'
> [  7%] Completed 'hwloc'
> [  7%] Built target hwloc
> make: *** [Makefile:130: all] Error 2
> ```

This seems to be fine. There is probably an error before that.

**Comment by matthiasdiener** [comment]
> In the previous (regular) build system, the directory structure of tests/benchmarks/examples was copied inside with files being symlinked to the actual files. However, this is missing in the cmake build system support.
> 
> I think it is worth adding this in order to maintain consistency with the previous build system and also to allow users/developers to work with tests/benchmarks/examples from multiple build-archs at the same time.

Should be fixed now.

**Comment by nitbhat** [comment]
That's right. The error was:
```
/ui/cwi/nitin/software/charm/src/arch/ucx/machine.C:20:10: fatal error: ucp/api/ucp.h: No such file or directory
 #include <ucp/api/ucp.h>
          ^~~~~~~~~~~~~~~
compilation terminated.
Fatal Error by charmc in directory /ui/cwi/nitin/software/charm/ucx-linux-x86_64-cmake/src/conv-core
   Command g++ -DCMK_GFORTRAN -I/ui/cwi/nitin/software/charm/ucx-linux-x86_64-cmake/bin/../include -I/usr/include/ -I./proc_management/ -I./proc_management/simple_pmi/ -D__CHARMC__=1 -DAMPI_ERRHANDLER_RETURN=0 -I/ui/cwi/nitin/software/charm/src/conv-core/../arch/util -DNDEBUG -O3 -std=gnu++11 -O2 -fno-stack-protector -fno-lifetime-dse -std=c++11 -c /ui/cwi/nitin/software/charm/src/arch/ucx/machine.C -o CMakeFiles/conv-machine.dir/__/arch/ucx/machine.C.o returned error code 1
charmc exiting...
```

My build command was wrong, I had to pass the UCX directory with `--basedir`. But it looks like even with `--basedir`, the cmake system doesn't use `--basedir`. I'm seeing the same error as above with `./buildcmake LIBS ucx-linux-x86_64 --enable-error-checking --suffix=cmake --with-production --basedir=/ui/cwi/nitin/software/ucx-1.5.1/build -j16 |& tee result`

**Comment by matthiasdiener** [comment]
> That's right. The error was:
> 
> ```
> /ui/cwi/nitin/software/charm/src/arch/ucx/machine.C:20:10: fatal error: ucp/api/ucp.h: No such file or directory
>  #include <ucp/api/ucp.h>
>           ^~~~~~~~~~~~~~~
> compilation terminated.
> Fatal Error by charmc in directory /ui/cwi/nitin/software/charm/ucx-linux-x86_64-cmake/src/conv-core
>    Command g++ -DCMK_GFORTRAN -I/ui/cwi/nitin/software/charm/ucx-linux-x86_64-cmake/bin/../include -I/usr/include/ -I./proc_management/ -I./proc_management/simple_pmi/ -D__CHARMC__=1 -DAMPI_ERRHANDLER_RETURN=0 -I/ui/cwi/nitin/software/charm/src/conv-core/../arch/util -DNDEBUG -O3 -std=gnu++11 -O2 -fno-stack-protector -fno-lifetime-dse -std=c++11 -c /ui/cwi/nitin/software/charm/src/arch/ucx/machine.C -o CMakeFiles/conv-machine.dir/__/arch/ucx/machine.C.o returned error code 1
> charmc exiting...
> ```
> 
> My build command was wrong, I had to pass the UCX directory with `--basedir`. But it looks like even with `--basedir`, the cmake system doesn't use `--basedir`. I'm seeing the same error as above with `./buildcmake LIBS ucx-linux-x86_64 --enable-error-checking --suffix=cmake --with-production --basedir=/ui/cwi/nitin/software/ucx-1.5.1/build -j16 |& tee result`

Yeah, `ucx` isn't supported yet, unfortunately.

**Comment by nitbhat** [comment]
I tested ofi build on bridges, and I see that the regular build works (`./build LIBS ofi-linux-x86_64 smp --with-production --suffix=regular --enable-error-checking -j16`). 


The cmake build (`./buildcmake LIBS ofi-linux-x86_64 smp --with-production --suffix=cmake --enable-error-checking -j16`) fails with the following error:

```
-- Performing Test CMK_DLL_USE_WIN32 - Failed
-- Performing Test CMK_TLS_SWITCHING_X86_64
-- Performing Test CMK_TLS_SWITCHING_X86_64 - Success
-- Performing Test CMK_TLS_SWITCHING_X86
-- Performing Test CMK_TLS_SWITCHING_X86 - Failed
-- Performing Test CMK_BALANCED_INJECTION_API
-- Performing Test CMK_BALANCED_INJECTION_API - Failed
-- Performing Test CMK_BUILD_ON_OFI
-- Performing Test CMK_BUILD_ON_OFI - Failed
CMake Error at cmake/detect-features.cmake:474 (message):
  Unable to build ofi.
Call Stack (most recent call first):
  CMakeLists.txt:264 (include)


-- Configuring incomplete, errors occurred!
See also "/home/nbhat4/software/charm_new/ofi-linux-x86_64-cmake/CMakeFiles/CMakeOutput.log".
See also "/home/nbhat4/software/charm_new/ofi-linux-x86_64-cmake/CMakeFiles/CMakeError.log".
```

The error is because of one of the configure tests not finding `fi_getinfo`. 
It looks like some of the include and library paths are not being added in the cmake system causing the build system to not find some libraries.

**Comment by matthiasdiener** [comment]
> I tested ofi build on bridges, and I see that the regular build works (`./build LIBS ofi-linux-x86_64 smp --with-production --suffix=regular --enable-error-checking -j16`).
> 
> The cmake build (`./buildcmake LIBS ofi-linux-x86_64 smp --with-production --suffix=cmake --enable-error-checking -j16`) fails with the following error:
> 
> ```
> -- Performing Test CMK_DLL_USE_WIN32 - Failed
> -- Performing Test CMK_TLS_SWITCHING_X86_64
> -- Performing Test CMK_TLS_SWITCHING_X86_64 - Success
> -- Performing Test CMK_TLS_SWITCHING_X86
> -- Performing Test CMK_TLS_SWITCHING_X86 - Failed
> -- Performing Test CMK_BALANCED_INJECTION_API
> -- Performing Test CMK_BALANCED_INJECTION_API - Failed
> -- Performing Test CMK_BUILD_ON_OFI
> -- Performing Test CMK_BUILD_ON_OFI - Failed
> CMake Error at cmake/detect-features.cmake:474 (message):
>   Unable to build ofi.
> Call Stack (most recent call first):
>   CMakeLists.txt:264 (include)
> 
> 
> -- Configuring incomplete, errors occurred!
> See also "/home/nbhat4/software/charm_new/ofi-linux-x86_64-cmake/CMakeFiles/CMakeOutput.log".
> See also "/home/nbhat4/software/charm_new/ofi-linux-x86_64-cmake/CMakeFiles/CMakeError.log".
> ```
> 
> The error is because of one of the configure tests not finding `fi_getinfo`.
> It looks like some of the include and library paths are not being added in the cmake system causing the build system to not find some libraries.

OFI build should be fixed now.

**Comment by matthiasdiener** [comment]
The LIBS target also works now. This should be ready to go.

**Comment by matthiasdiener** [comment]
The LIBS target also works now. This should be ready to go.
